examples/Backtesting/transfer_learning.py

```diff
@@ -0,0 +1,120 @@
+
+N_MC_ITERATIONS = 2
+N_DOE_ITERATIONS = 4
+
+#### Defining the tasks
+
+# We use two similar two-dimensional test functions imported from BoTorch.
+# We use them to create a lookup table after creating the search space and the objective.
+three_hump_camel = ThreeHumpCamel()
+six_hump_camel = SixHumpCamel()
+
+#### Creating the searchspace and the objective
+
+# The parameter `POINTS_PER_DIM` controls the number of points per dimension.
+# Note that the searchspace will have `POINTS_PER_DIM**2` many points.
+
+BOUNDS = three_hump_camel.bounds
+POINTS_PER_DIM = 20
+
+x1 = NumericalDiscreteParameter(
+    name="x_1", values=list(np.linspace(BOUNDS[0, 0], BOUNDS[1, 0], POINTS_PER_DIM))
+)
+x2 = NumericalDiscreteParameter(
+    name="x_2", values=list(np.linspace(BOUNDS[0, 1], BOUNDS[1, 1], POINTS_PER_DIM))
+)
+
+# We now append a ``TaskParameter``. This parameter contains the information about the
+# different tasks that we have in this transfer learning example.
+task_param = TaskParameter(name="Function", values=("ThreeHump", "SixHump"))
+parameters = [x1, x2, task_param]
+
+# We now create the searchspace, objective and campaign.
+
+searchspace = SearchSpace.from_product(parameters=parameters)
+objective = Objective(
+    mode="SINGLE", targets=[NumericalTarget(name="Target", mode="MIN")]
+)
+campaign = Campaign(searchspace=searchspace, objective=objective)
+
+#### Generating the lookup table
+
+# The lookup table for the parameters and their values need to contain one column per
+# parameter, including one for the TaskParameter. The values in that column describe
+# for which task a measurement was made.
+
+# Generate the grid coordinates using numpy
+x = parameters[0].values
+y = parameters[1].values
+xx, yy = np.meshgrid(x, y)
+
+lookup = pd.DataFrame({"x_1": xx.flatten(), "x_2": yy.flatten()})
+
+# We randomly decide for each row which function we measure
+lookup["Function"] = np.random.choice(["ThreeHump", "SixHump"], lookup.shape[0])
+
+
+# This is a little helper function that is used to apply the function three_hump_camel
+# resp. six_humP-camel if the value of the "Function" column is ``"ThreeHump"`` resp.
+# ``"SixHump"``.
+def _choose_task(x1, x2, function):
+    func_dict = {"ThreeHump": three_hump_camel, "SixHump": six_hump_camel}
+    return float(func_dict[function](Tensor((x1, x2))))
+
+
+# Apply the helper function to the lookup dataframe
+lookup["Target"] = lookup.apply(
+    lambda x: _choose_task(x["x_1"], x["x_2"], x["Function"]), axis=1
+)
+
+# We print the first ten rows to show how the table looks like
+print(lookup.head(10))
+
+
+#### Performing the simulation loop
+
+# We can now use the `simulate_transfer_learning` function to simulate a full experiment.
+
+results = simulate_transfer_learning(
+    campaign,
+    lookup,
+    batch_quantity=1,
+    n_doe_iterations=N_DOE_ITERATIONS,
+    n_mc_iterations=N_MC_ITERATIONS,
+)
+
```

Comment:
theres no comparison to just plain simulating the campaigns without transfer learning which feels kind of esssential for such a demo?

Reply:
Such a comparison is included now. This is the output with the current configuration:
![image](https://github.com/emdgroup/baybe/assets/84016733/48f97375-ff0f-4c30-911e-8eada4109c89)
It is the best I could come up with while keeping the runtime reasonable. If this is acceptable, please resolve.

Reply:
not sure why youre so stingy on the MC runs, this takes maybe 45min (and n iteraions cans till be reduced):
![image](https://github.com/emdgroup/baybe/assets/17951239/c39c1ccd-9ecd-4eb7-9560-2c7c138d88e5)

this actually doesnt look that great for transfer learning, are we sure everything is correct?

Reply:
I wanted to keep number of MC iterations low since examples are tested/executed in our tox pipelines if you do not configure them to be ignored. So I wanted to keep the overall execution time below 1min.
I do not know. Maybe the example is not very well suited for a Transfer Learning example?

Reply:
I was not talking about increasing the number of mc runs in the published example, but just for us to check. did you not want to check whether this example actually makes sense and produces a reasonable result? it can reamin with n_mc=2 for the published version

Reply:
i thought these functions were specifically designed for a transfer learning use case?

Reply:
Well, I had a look at the available test functions in botorch and these were the ones that seemed to make the most sense for me, given that I wanted to have a low-dimensional problem. I will however test it again with some higher dimensional variations of the Hartmann function - there is an unconstrained version as well as two constrained versions. This might also be a nice use case for active values: After having learned the unconstrained function and a constrained variant, learn the other constrained variant.

Reply:
not sure thats the right approach, the transfer learning will only yield better results if the two examples chosen have some common behavior. Is that in any way indicated for the hump functions used there or the hartmannX functions?
might be required to make some custom examples (maybe get inspiration form some neurips papers, there might be some that used analytical examples, otherwise... we also have the public reaction data from the benchmark)

Reply:
These produce quite nice plots in a reasonable time:
![image](https://github.com/emdgroup/baybe/assets/84016733/91572c48-d412-4ea8-a542-b069ea45ed12)
So I'd propose to use these for the example here.

Reply:
I have changed the example accordingly, this is now the latest plot:
![image](https://github.com/emdgroup/baybe/assets/84016733/705c04b8-24a9-4b2a-ba29-3208678e24c0)
Please resolve if this is gucci.

Reply:
1) did you make changes to the botorch version? Locally on your branch Im getting import error indicating there have been changes in botorch verisons regarding these functions (im on 0.8.5)

2) Please reduce the number of doe and mc iterations to the numbers used in other examples

Reply:
I had to re-install my environment, then it worked.

Reply:
so with what botorch versions does this work? it needs to be correct in pyproject.toml

Reply:
According to changelog, it's 0.9.3. Will add.

Reply:
Since we did already discuss this issue, I'll resolve here and create an additional comment to avoid accidental merging.